### PR TITLE
CRUD API for addons 

### PR DESCRIPTION
**Pull request by multi-io**
**What this PR does / why we need it**:

This adds endpoints to the API so users can C/R/U/D addons in their own clusters. We also added an option for passing a whitelist of the actually accessible addons to the apiserver, so you can prevent users from uninstalling canal or other vital addons.

**Special notes for your reviewer**:

Open issues:

- unit tests

- using runtime.RawExtension in the AddonSpec API type -- is this OK?

```release-note
added API endpoints for user to create/read/update/delete addons in the clusters
```

### CODE REVIEW

File: api/cmd/kubermatic-api/options.go [hunk 1]
```diff
@@ -90,6 +94,12 @@ func newServerRunOptions() (serverRunOptions, error) {
 	default:
 		return s, fmt.Errorf("--expose-strategy must be either `NodePort` or `LoadBalancer`, got %q", rawExposeStrategy)
 	}
+
+	s.accessibleAddons = map[string]bool{}
```
Comment (alvaroaleman) [inline api/cmd/kubermatic-api/options.go]: Nit: Using `k8s.io/apimachinery/pkg/util/sets.String` makes these things much more readable :)

File: config/kubermatic/values.yaml [hunk 2]
```diff
@@ -128,6 +128,10 @@ kubermatic:
 
   api:
     replicas: 2
+    accessibleAddons:
+    - dashboard
```
Comment (maciaszczykm) [inline config/kubermatic/values.yaml]: I think we shouldn't list Dashboard at least until connect window show instructions how to connect to it. Disabling the addon will not make these intructions disappear and can confuse users.

Zrzut ekranu 2019-07-18 o 11 43 37

File: api/pkg/api/v1/types.go [hunk 3]
```diff
@@ -2,6 +2,7 @@ package v1
 
 import (
 	"encoding/json"
+	"k8s.io/apimachinery/pkg/runtime"
```
Comment (maciaszczykm) [inline api/pkg/api/v1/types.go]: The new line is missing.

File: api/cmd/kubermatic-api/options.go (hunk 1 above)
Comment (multi-io) [reply api/cmd/kubermatic-api/options.go]: You're right, didn't think of that. I'll change it.

**Comment by multi-io** [comment]
I'll rebase, address the issues you guys brought up and add some tests too because we're actually using this in our fork currently only. Will report back early next week.

**Comment by multi-io** [comment]
PR updated:
- unit tests for the crud operations
- map[string]bool replaced w/ string set
- use map[string]interface{} instead of RawExtension for the public-facing AddonSpec.Variables field
- empty default accessibleAddons list

**Comment by maciaszczykm** [comment]
No further remarks from my side.

/approve

**Comment by multi-io** [comment]
Rebased. Adapted to the dynamic-datacenters/CRDs stuff, should work but couldn't test it all since we don't currently have a stock Kubermatic installation (and our fork is still using the static datacenters.yaml/multi-seed kubeconfig). So basically if the providers initialization in main.go works, the CRUD apis themselves are covered by the unit tests.

**Comment by alvaroaleman** [comment]
> Rebased. Adapted to the dynamic-datacenters/CRDs stuff, should work but couldn't test it all since we don't currently have a stock Kubermatic installation (and our fork is still using the static datacenters.yaml/multi-seed kubeconfig). So basically if the providers initialization in main.go works, the CRUD apis themselves are covered by the unit tests.

So far the old `datacenters.yaml` continues to work and will for some time, the feature is not yet ready. We just migrate it internally already to a new data structure that better represents the fact that seeds and datacenters are a complete distinct thing :)

/test all

File: api/pkg/controller/rbac/sync_resource.go [hunk 4]
```diff
@@ -301,3 +310,130 @@ func shouldSkipRBACRoleBindingForNamedResource(projectName string, objectGVR sch
 	}
 	return false, generatedRole, nil
 }
+
+func (c *resourcesController) ensureRBACRoleForClusterAddons(projectName string, object metav1.Object, clusterProvider *ClusterProvider) error {
+	cluster, ok := object.(*kubermaticv1.Cluster)
+	if !ok {
+		panic("ensureRBACRoleBindingForClusterAddons called with non-cluster")
```
Comment (alvaroaleman) [inline api/pkg/controller/rbac/sync_resource.go]: Err. No way.

File: api/pkg/controller/rbac/sync_resource.go [hunk 5]
```diff
@@ -301,3 +310,130 @@ func shouldSkipRBACRoleBindingForNamedResource(projectName string, objectGVR sch
...
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *resourcesController) ensureRBACRoleBindingForClusterAddons(projectName string, object metav1.Object, clusterProvider *ClusterProvider) error {
+	cluster, ok := object.(*kubermaticv1.Cluster)
+	if !ok {
+		panic("ensureRBACRoleBindingForClusterAddons called with non-cluster")
```
Comment (alvaroaleman) [inline api/pkg/controller/rbac/sync_resource.go]: same. Just return an error.

Comment (alvaroaleman) [reply api/pkg/controller/rbac/sync_resource.go]: (the fact that this should never happen is absolutely not a reason to put a panic in production code)

Comment (multi-io) [reply api/pkg/controller/rbac/sync_resource.go]: I'll change it. I'm pretty sure I just copied it from somewhere though.

**Comment by multi-io** [comment]
Force-push  with just the two panics replaced w/ error returns.

File: api/pkg/controller/rbac/sync_resource.go (hunk 5 above)
Comment (multi-io) [reply api/pkg/controller/rbac/sync_resource.go]: https://github.com/kubermatic/kubermatic/blob/05b9ef7569ad1067678d165a4fe07c61f6760316/api/pkg/controller/rbac/informer_provider.go#L49 was the place I think.

**Comment by alvaroaleman** [comment]
/test all
@zreigz can you lgtm if this looks good?

**Pull request merged**